### PR TITLE
Private Endpoint for App Service

### PR DESCRIPTION
**Pull request by hudua**
## Overview/Summary

Added optional support for App Service (Linux container) to enable private endpoint for inbound traffic. This provides users the option of deploying machine learning as a public API or as a private API which is accessible only using private endpoint.

This resolves issue #131.

## This PR fixes/adds/changes/removes

1. Machine learning landing zone documentation updated to reflect addition of app service private endpoint
2. Machine learning landing zone bicep scripts updated to support optional private endpoint for app service
3. Updated schema definition and change log
4. Updated existing deployment JSON files
5. Ran schema tests/schemas/run-tests.sh test
6. Updated ad-hoc deployment tests for app service under tests/landingzones/lz-machinelearning/deployment-tests

### Breaking Changes

1. Introduced new `enablePrivateEndpoint` field for AppServiceLinuxContainer config file

## Testing Evidence

App service deployed with private endpoint, when `enablePrivateEndpoint` is true

Screenshot 2022-01-17 002454

Screenshot 2022-01-17 002510

App service deployed with private endpoint, when `enablePrivateEndpoint` is false

Screenshot 2022-01-17 003216

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/azure/CanadaPubSecALZ/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/azure/CanadaPubSecALZ/issues)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/azure/CanadaPubSecALZ/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.

### CODE REVIEW

**Comment by hudua** [comment]
Will kindly ask @SenthuranSivananthan to deploy e2e and I will run through deployment tests to verify:

- private inbound endpoint
  - With private endpoint enabled, I have currently left vnet integration enabled. Will test as such, and will manually disable vnet integration to see if image from ACR with private endpoint can still be pulled.
* public inbound endpoint

File: docs/archetypes/machinelearning.md [hunk 1]
```diff
@@ -131,7 +131,7 @@ Once the machine learning archetype is deployed and available to use, access con
 | Azure Storage Account for Azure ML | Network ACL deny | Private endpoint on `blob`, `file` + DNS registration to either hub or spoke | `privateEndpoints`|
 | Azure Data Factory | Public network access disabled, Azure integration runtime with managed virtual network | Private endpoint on `dataFactory` + DNS registration to either hub or spoke | `privateEndpoints`|
 | Azure Kubernetes Service | Private cluster, network profile set with either kubenet or Azure CNI | N/A | `aks`|
-| Azure App Service | Virtual Network integration | N/A | `appService` | 
+| Azure App Service | Virtual Network integration. Public network access can be disabled, using private endpoint instead | Private endpoint on `azureWebsites` + DNS registration to either hub or spoke | `appService`, `privateEndpoints` |
```
Comment (SenthuranSivananthan) [inline docs/archetypes/machinelearning.md:134]: Let's add a note & reference to Azure Docs that Private Endpoint support requires `Premium`

https://docs.microsoft.com/en-us/azure/app-service/networking/private-endpoint

**Review by SenthuranSivananthan — CHANGES_REQUESTED** [review]
Thanks for the great addition.  2 notes:

1. Let's update the docs to state that Private Endpoint support requires Premium SKU
2. I've setup the environment for you to run E2E testing.  Please share the results when ready.

File: docs/archetypes/machinelearning.md (hunk 1 above)
Comment (hudua) [reply docs/archetypes/machinelearning.md:134]: Sounds good - will do that now!

**Comment by hudua** [comment]
Tested the following...

Private endpoint is enabled
Screenshot 2022-01-21 110928

Requesting prediction inside vnet through private endpoint
image

Not able when outside vnet
image
image

**Review by SenthuranSivananthan — APPROVED** [review]
Approved.  Thanks @hudua

**Pull request merged**